--- gdb-7.8.orig/gdb/rs6000-tdep.c	2014-06-11 11:34:41.000000000 -0500
+++ gdb-7.8/gdb/rs6000-tdep.c	2014-09-12 14:59:16.795425453 -0500
@@ -3191,9 +3191,14 @@ rs6000_frame_cache (struct frame_info *t
     }
 
   if (!fdata.frameless)
-    /* Frameless really means stackless.  */
-    cache->base
-      = read_memory_unsigned_integer (cache->base, wordsize, byte_order);
+    {
+      /* Frameless really means stackless.  */
+      LONGEST backchain;
+
+      if (safe_read_memory_integer (cache->base, wordsize,
+				    byte_order, &backchain))
+        cache->base = (CORE_ADDR) backchain;
+    }
 
   trad_frame_set_value (cache->saved_regs,
 			gdbarch_sp_regnum (gdbarch), cache->base);
diff -uprN gdb-7.8.orig/gdb/testsuite/gdb.arch/powerpc-stackless.exp gdb-7.8/gdb/testsuite/gdb.arch/powerpc-stackless.exp
--- gdb-7.8.orig/gdb/testsuite/gdb.arch/powerpc-stackless.exp	1969-12-31 18:00:00.000000000 -0600
+++ gdb-7.8/gdb/testsuite/gdb.arch/powerpc-stackless.exp	2014-09-12 14:59:16.795425453 -0500
@@ -0,0 +1,42 @@
+# Copyright 2014 Free Software Foundation, Inc.
+
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+# Testcase for PR tdep/17379.
+
+if {![istarget "powerpc*-*-*"]} then {
+    verbose "Skipping powerpc-stackless.exp"
+    return
+}
+
+standard_testfile .S
+
+if { [prepare_for_testing $testfile.exp $testfile $srcfile] } {
+    untested powerpc-stackless.exp
+    return -1
+}
+
+# Run until SIGSEGV.
+gdb_run_cmd
+
+set test "run until SIGSEGV"
+gdb_test_multiple "" $test {
+    -re "Program received signal SIGSEGV.*$gdb_prompt $" {
+  pass $test
+    }
+}
+
+# Ensure that 'info registers' works properly and does not generate
+# an internal-error.
+gdb_test "info registers" "r0.*" "info registers"
diff -uprN gdb-7.8.orig/gdb/testsuite/gdb.arch/powerpc-stackless.S gdb-7.8/gdb/testsuite/gdb.arch/powerpc-stackless.S
--- gdb-7.8.orig/gdb/testsuite/gdb.arch/powerpc-stackless.S	1969-12-31 18:00:00.000000000 -0600
+++ gdb-7.8/gdb/testsuite/gdb.arch/powerpc-stackless.S	2014-09-12 14:59:16.795425453 -0500
@@ -0,0 +1,24 @@
+/* This testcase is part of GDB, the GNU debugger.
+
+   Copyright 2014 Free Software Foundation, Inc.
+
+   This program is free software; you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation; either version 3 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.  */
+
+#include <ppc-asm.h>
+
+FUNC_START(main)
+        li      sp,0
+        mtlr    sp
+        blr
+FUNC_END(main)
